File: backend/app/schemas/bank_reconciliation.py
"""
Bank Reconciliation Pydantic Schemas
For API request/response validation
"""

from pydantic import BaseModel, Field, ConfigDict
from datetime import date
from decimal import Decimal
from typing import Optional, List, Dict, Any
from uuid import UUID

from app.schemas.base import AuditableSchema, CompanyBaseSchema


# ============ Bank Account Schemas ============

class BankAccountBase(CompanyBaseSchema):
    """Base schema for bank account"""
    name: str = Field(..., min_length=1, max_length=100)
    account_number: str = Field(..., min_length=1, max_length=50)
    bank_name: str = Field(..., min_length=1, max_length=100)
    branch: Optional[str] = Field(None, max_length=100)
    iban: Optional[str] = Field(None, max_length=50)
    currency: str = Field(default="PKR", min_length=3, max_length=3)
    opening_balance: Decimal = Field(default=0, ge=0)
    current_balance: Decimal = Field(default=0)
    is_active: bool = True


class BankAccountCreate(BankAccountBase):
    """Schema for creating bank account"""
    pass


class BankAccountUpdate(BaseModel):
    """Schema for updating bank account"""
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    account_number: Optional[str] = Field(None, min_length=1, max_length=50)
    bank_name: Optional[str] = Field(None, min_length=1, max_length=100)
    branch: Optional[str] = Field(None, max_length=100)
    iban: Optional[str] = Field(None, max_length=50)
    currency: Optional[str] = Field(None, min_length=3, max_length=3)
    opening_balance: Optional[Decimal] = Field(None, ge=0)
    current_balance: Optional[Decimal] = None
    is_active: Optional[bool] = None


class BankAccountResponse(AuditableSchema):
    """Schema for bank account response"""
    model_config = ConfigDict(from_attributes=True)
    
    id: UUID
    name: str
    account_number: str
    bank_name: str
    branch: Optional[str]
    iban: Optional[str]
    currency: str
    opening_balance: Decimal
    current_balance: Decimal
    is_active: bool


# ============ Bank Statement Schemas ============

class BankStatementImport(CompanyBaseSchema):
    """Schema for importing bank statement CSV"""
    bank_account_id: UUID
    statement_date: date
    opening_balance: Decimal
    closing_balance: Decimal
    transactions: List[Dict[str, Any]]


class BankStatementResponse(BaseModel):
    """Schema for bank statement response"""
    model_config = ConfigDict(from_attributes=True)
    
    id: UUID
    bank_account_id: UUID
    statement_date: date
    opening_balance: Decimal
    closing_balance: Decimal
    transactions_count: int
    imported_at: date
    imported_by: UUID


# ============ Reconciliation Session Schemas ============

class ReconciliationSessionStart(CompanyBaseSchema):
    """Schema for starting reconciliation session"""
    bank_account_id: UUID
    period_month: int = Field(..., ge=1, le=12)
    period_year: int = Field(..., ge=2020)
    opening_balance: Decimal
    closing_balance_per_bank: Decimal


class ReconciliationSessionMatch(BaseModel):
    """Schema for matching transactions"""
    bank_transaction_id: str
    system_transaction_id: UUID
    match_type: str = Field(default="manual", pattern="^(auto|manual)$")


class ReconciliationSessionComplete(BaseModel):
    """Schema for completing reconciliation"""
    notes: Optional[str] = None


class ReconciliationSessionResponse(AuditableSchema):
    """Schema for reconciliation session response"""
    model_config = ConfigDict(from_attributes=True)
    
    id: UUID
    bank_account_id: UUID
    period_month: int
    period_year: int
    opening_balance: Decimal
    closing_balance_per_bank: Decimal
    closing_balance_per_books: Decimal
    difference: Decimal
    status: str
    completed_at: Optional[date]
    completed_by: Optional[UUID]


class ReconciliationSessionDetail(ReconciliationSessionResponse):
    """Reconciliation session with transactions"""
    bank_transactions: Optional[List[Dict[str, Any]]] = None
    system_transactions: Optional[List[Dict[str, Any]]] = None
    matched_transactions: Optional[List[Dict[str, Any]]] = None


# ============ PDC Schemas ============

class PDCBase(CompanyBaseSchema):
    """Base schema for PDC"""
    cheque_number: str = Field(..., min_length=1, max_length=20)
    bank_name: str = Field(..., min_length=1, max_length=100)
    cheque_date: date
    amount: Decimal = Field(..., gt=0)
    party_type: str = Field(..., pattern="^(customer|vendor)$")
    party_id: UUID
    status: str = Field(default="pending", pattern="^(pending|deposited|cleared|bounced|returned)$")


class PDCCreate(PDCBase):
    """Schema for creating PDC"""
    pass


class PDCUpdateStatus(BaseModel):
    """Schema for updating PDC status"""
    status: str = Field(..., pattern="^(pending|deposited|cleared|bounced|returned)$")
    bounce_reason: Optional[str] = Field(None, max_length=200)


class PDCResponse(BaseModel):
    """Schema for PDC response"""
    model_config = ConfigDict(from_attributes=True)
    
    id: UUID
    cheque_number: str
    bank_name: str
    cheque_date: date
    amount: Decimal
    party_type: str
    party_id: UUID
    status: str
    deposited_at: Optional[date]
    cleared_at: Optional[date]
    bounced_at: Optional[date]
    bounce_reason: Optional[str]
    payment_id: Optional[UUID]
    created_at: date
    updated_at: date


class PDCMaturityItem(BaseModel):
    """PDC maturity report item"""
    id: UUID
    cheque_number: str
    party_name: str
    party_type: str
    amount: Decimal
    cheque_date: date
    days_until_maturity: int
    status: str


# ============ Cash Position Schema ============

class CashPositionSummary(BaseModel):
    """Cash position summary"""
    total_cash_in_hand: Decimal
    total_cash_at_bank: Decimal
    total_pdc_receivable: Decimal
    total_pdc_payable: Decimal
    net_cash_position: Decimal
    bank_accounts: List[Dict[str, Any]]

File: frontend/src/lib/api/bank-reconciliation.ts
/**
 * Bank Reconciliation API Client
 * Handles all API calls to the Bank Reconciliation backend
 * Pakistani banking conventions: PKR currency, DD/MM/YYYY dates
 */

import api from '@/lib/api';
import axios from 'axios';

const API_BASE = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000/api';

function getAuthToken(): string | null {
  return typeof window !== 'undefined' ? localStorage.getItem('access_token') : null;
}

function authHeaders(): Record<string, string> {
  const token = getAuthToken();
  if (!token) throw new Error('Authentication required');
  return { Authorization: `Bearer ${token}` };
}

// ============ TypeScript Types ============

export interface BankAccount {
  id: string;
  name: string;
  account_number: string;
  bank_name: string;
  branch?: string;
  iban?: string;
  currency: string;
  opening_balance: number;
  current_balance: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface BankStatementTransaction {
  date: string;
  description: string;
  debit: number;
  credit: number;
  balance: number;
  cheque_number?: string;
}

export interface BankStatementImportResponse {
  id: string;
  bank_account_id: string;
  statement_date: string;
  opening_balance: number;
  closing_balance: number;
  transactions_count: number;
  imported_at: string;
  imported_by: string;
}

export interface ReconciliationSession {
  id: string;
  bank_account_id: string;
  period_month: number;
  period_year: number;
  opening_balance: number;
  closing_balance_per_bank: number;
  closing_balance_per_books: number;
  difference: number;
  status: 'in_progress' | 'completed' | 'cancelled';
  completed_at?: string;
  completed_by?: string;
  created_at: string;
  updated_at: string;
}

export interface ReconciliationSessionDetail extends ReconciliationSession {
  bank_transactions: BankTransaction[];
  system_transactions: SystemTransaction[];
  matched_transactions: MatchedTransaction[];
}

export interface BankTransaction {
  id: string;
  date: string;
  description: string;
  debit: number;
  credit: number;
  balance: number;
  cheque_number?: string;
  reference?: string;
  matched: boolean;
}

export interface SystemTransaction {
  id: string;
  date: string;
  description: string;
  amount: number;
  type: 'payment' | 'receipt' | 'journal';
  reference?: string;
  party_name?: string;
  matched: boolean;
}

export interface MatchedTransaction {
  bank_transaction_id: string;
  system_transaction_id: string;
  match_type: 'auto' | 'manual';
  matched_at: string;
}

export interface MatchingSuggestion {
  auto_matches: AutoMatch[];
  unmatched_bank_transactions: BankTransaction[];
  unmatched_system_transactions: SystemTransaction[];
  matching_rules_applied: string[];
}

export interface AutoMatch {
  bank_transaction: BankTransaction;
  system_transaction: SystemTransaction;
  confidence: 'high' | 'medium' | 'low';
  match_reason: string;
}

export interface ReconciliationStatement {
  bank_account: {
    id: string;
    name: string;
    bank_name: string;
    account_number: string;
  };
  period: {
    from_date: string | null;
    to_date: string | null;
  };
  reconciled_sessions: Array<{
    id: string;
    period: string;
    opening_balance: number;
    closing_balance_per_bank: number;
    closing_balance_per_books: number;
    difference: number;
    completed_at: string | null;
  }>;
  summary: {
    total_sessions: number;
    last_reconciled: string | null;
    last_difference: number;
  };
}

// ============ PDC Types ============

export type PDCStatus = 'pending' | 'deposited' | 'cleared' | 'bounced' | 'returned';
export type PDCPartyType = 'customer' | 'vendor';

export interface PDC {
  id: string;
  cheque_number: string;
  bank_name: string;
  cheque_date: string;
  amount: number;
  party_type: PDCPartyType;
  party_id: string;
  party_name?: string;
  status: PDCStatus;
  deposited_at?: string;
  cleared_at?: string;
  bounced_at?: string;
  bounce_reason?: string;
  payment_id?: string;
  created_at: string;
  updated_at: string;
}

export interface PDCCreateData {
  cheque_number: string;
  bank_name: string;
  cheque_date: string;
  amount: number;
  party_type: PDCPartyType;
  party_id: string;
}

export interface PDCMaturityItem {
  id: string;
  cheque_number: string;
  party_name: string;
  party_type: PDCPartyType;
  amount: number;
  cheque_date: string;
  days_until_maturity: number;
  status: PDCStatus;
}

export interface CashPositionSummary {
  total_cash_in_hand: number;
  total_cash_at_bank: number;
  total_pdc_receivable: number;
  total_pdc_payable: number;
  net_cash_position: number;
  bank_accounts: Array<{
    id: string;
    name: string;
    bank_name: string;
    account_number: string;
    balance: number;
  }>;
}

// ============ CSV Import Types ============

export interface CSVColumnMapping {
  date?: string;
  description?: string;
  debit?: string;
  credit?: string;
  balance?: string;
  cheque_number?: string;
  reference?: string;
}

export interface CSVParseResult {
  headers: string[];
  rows: Record<string, string>[];
  rowCount: number;
}

// ============ Bank Accounts ============

export async function fetchBankAccounts(isActive = true): Promise<BankAccount[]> {
  const response = await api.get(`/api/bank-reconciliation/accounts`, {
    params: { is_active: isActive },
  });
  return response.data;
}

export async function createBankAccount(data: {
  name: string;
  account_number: string;
  bank_name: string;
  branch?: string;
  iban?: string;
  currency?: string;
  opening_balance?: number;
  current_balance?: number;
  is_active?: boolean;
}): Promise<BankAccount> {
  const response = await api.post(`/api/bank-reconciliation/accounts`, data);
  return response.data;
}

export async function updateBankAccount(
  id: string,
  data: Partial<{
    name: string;
    account_number: string;
    bank_name: string;
    branch?: string;
    iban?: string;
    currency?: string;
    opening_balance?: number;
    current_balance?: number;
    is_active?: boolean;
  }>
): Promise<BankAccount> {
  const response = await api.put(`/api/bank-reconciliation/accounts/${id}`, data);
  return response.data;
}

// ============ Bank Statement Import ============

/**
 * Parse CSV file client-side to extract headers and preview rows
 */
export function parseCSVFile(file: File): Promise<CSVParseResult> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onload = (e) => {
      try {
        const text = e.target?.result as string;
        const lines = text.split('\n').filter((line) => line.trim());
        if (lines.length === 0) {
          reject(new Error('CSV file is empty'));
          return;
        }

        // Parse headers
        const headers = lines[0].split(',').map((h) => h.trim().replace(/^"|"$/g, ''));

        // Parse rows
        const rows: Record<string, string>[] = [];
        for (let i = 1; i < lines.length; i++) {
          const values = lines[i].split(',').map((v) => v.trim().replace(/^"|"$/g, ''));
          const row: Record<string, string> = {};
          headers.forEach((header, idx) => {
            row[header] = values[idx] || '';
          });
          rows.push(row);
        }

        resolve({ headers, rows, rowCount: rows.length });
      } catch (err) {
        reject(err);
      }
    };
    reader.onerror = () => reject(new Error('Failed to read CSV file'));
    reader.readAsText(file);
  });
}

/**
 * Auto-detect column mapping from CSV headers
 */
export function autoDetectColumnMapping(headers: string[]): CSVColumnMapping {
  const mapping: CSVColumnMapping = {};
  const lowerHeaders = headers.map((h) => h.toLowerCase());

  // Date detection
  const dateIdx = lowerHeaders.findIndex(
    (h) => h === 'date' || h === 'transaction_date' || h === 'value_date' || h === 'posting_date'
  );
  if (dateIdx >= 0) mapping.date = headers[dateIdx];

  // Description detection
  const descIdx = lowerHeaders.findIndex(
    (h) =>
      h === 'description' ||
      h === 'narration' ||
      h === 'particulars' ||
      h === 'details' ||
      h === 'transaction_description'
  );
  if (descIdx >= 0) mapping.description = headers[descIdx];

  // Debit detection
  const debitIdx = lowerHeaders.findIndex(
    (h) => h === 'debit' || h === 'withdrawal' || h === 'dr' || h === 'amount_dr'
  );
  if (debitIdx >= 0) mapping.debit = headers[debitIdx];

  // Credit detection
  const creditIdx = lowerHeaders.findIndex(
    (h) => h === 'credit' || h === 'deposit' || h === 'cr' || h === 'amount_cr'
  );
  if (creditIdx >= 0) mapping.credit = headers[creditIdx];

  // Balance detection
  const balIdx = lowerHeaders.findIndex(
    (h) => h === 'balance' || h === 'running_balance' || h === 'closing_balance'
  );
  if (balIdx >= 0) mapping.balance = headers[balIdx];

  // Cheque number detection
  const chequeIdx = lowerHeaders.findIndex(
    (h) =>
      h === 'cheque_number' ||
      h === 'cheque_no' ||
      h === 'cheque_no.' ||
      h === 'instrument_no' ||
      h === 'ref_no'
  );
  if (chequeIdx >= 0) mapping.cheque_number = headers[chequeIdx];

  return mapping;
}

export async function importBankStatement(
  bankAccountId: string,
  statementDate: string,
  openingBalance: number,
  closingBalance: number,
  file: File
): Promise<BankStatementImportResponse> {
  const formData = new FormData();
  formData.append('bank_account_id', bankAccountId);
  formData.append('statement_date', statementDate);
  formData.append('opening_balance', openingBalance.toString());
  formData.append('closing_balance', closingBalance.toString());
  formData.append('file', file);

  const token = getAuthToken();
  if (!token) throw new Error('Authentication required');

  const response = await api.post(`/api/bank-reconciliation/import-statement`, formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
  return response.data;
}

// ============ Reconciliation Sessions ============

export async function fetchReconciliationSessions(
  bankAccountId?: string,
  status?: string
): Promise<ReconciliationSession[]> {
  const params: Record<string, string> = {};
  if (bankAccountId) params.bank_account_id = bankAccountId;
  if (status) params.status = status;

  const response = await api.get(`/api/bank-reconciliation/sessions`, { params });
  return response.data;
}

export async function fetchReconciliationSession(
  sessionId: string
): Promise<ReconciliationSessionDetail> {
  const response = await api.get(`/api/bank-reconciliation/sessions/${sessionId}`);
  return response.data;
}

export async function startReconciliationSession(data: {
  bank_account_id: string;
  period_month: number;
  period_year: number;
  opening_balance: number;
  closing_balance_per_bank: number;
}): Promise<ReconciliationSession> {
  const response = await api.post(`/api/bank-reconciliation/sessions`, data);
  return response.data;
}

export async function matchTransactions(
  sessionId: string,
  data: {
    bank_transaction_id: string;
    system_transaction_id: string;
    match_type?: 'auto' | 'manual';
  }
): Promise<{ message: string }> {
  const response = await api.post(
    `/api/bank-reconciliation/sessions/${sessionId}/match`,
    data
  );
  return response.data;
}

export async function completeReconciliation(sessionId: string): Promise<{
  id: string;
  status: string;
  completed_at: string;
}> {
  const response = await api.post(
    `/api/bank-reconciliation/sessions/${sessionId}/complete`,
    {}
  );
  return response.data;
}

// ============ Matching Suggestions ============

export async function fetchMatchingSuggestions(
  sessionId?: string,
  bankAccountId?: string
): Promise<MatchingSuggestion> {
  const params: Record<string, string> = {};
  if (sessionId) params.session_id = sessionId;
  if (bankAccountId) params.bank_account_id = bankAccountId;

  const response = await api.get(`/api/bank-reconciliation/matching-suggestions`, { params });
  return response.data;
}

// ============ Reports ============

export async function fetchReconciliationStatement(
  bankAccountId: string,
  fromDate?: string,
  toDate?: string
): Promise<ReconciliationStatement> {
  const params: Record<string, string> = { bank_account_id: bankAccountId };
  if (fromDate) params.from_date = fromDate;
  if (toDate) params.to_date = toDate;

  const response = await api.get(`/api/bank-reconciliation/reports/statement`, { params });
  return response.data;
}

// ============ PDCs ============

export async function fetchPDCs(
  status?: string,
  partyType?: string,
  fromDate?: string,
  toDate?: string
): Promise<PDC[]> {
  const params: Record<string, string> = {};
  if (status && status !== 'all') params.status = status;
  if (partyType && partyType !== 'all') params.party_type = partyType;
  if (fromDate) params.from_date = fromDate;
  if (toDate) params.to_date = toDate;

  const response = await api.get(`/api/bank-reconciliation/pdcs`, { params });
  return response.data;
}

export async function createPDC(data: PDCCreateData): Promise<PDC> {
  const response = await api.post(`/api/bank-reconciliation/pdcs`, data);
  return response.data;
}

export async function updatePDCStatus(
  id: string,
  status: PDCStatus,
  bounceReason?: string
): Promise<PDC> {
  const response = await api.put(`/api/bank-reconciliation/pdcs/${id}/status`, {
    status,
    bounce_reason: bounceReason,
  });
  return response.data;
}

export async function depositPDC(id: string): Promise<PDC> {
  const response = await api.post(`/api/bank-reconciliation/pdcs/${id}/deposit`, {});
  return response.data;
}

export async function fetchPDCMaturityReport(daysAhead = 30): Promise<PDCMaturityItem[]> {
  const response = await api.get(`/api/bank-reconciliation/pdcs/maturity-report`, {
    params: { days_ahead: daysAhead },
  });
  return response.data;
}

// ============ Cash Position ============

export async function fetchCashPosition(): Promise<CashPositionSummary> {
  const response = await api.get(`/api/bank-reconciliation/cash-position`);
  return response.data;
}
